feat: JSON Masking

jamfor352 opened the pull request:
Hi!

AKHQ is fantastic, but the data masking using regex doesn't really scale well if you have complex personal data (pii) requirements (for example, if you need to mask *all* fields by default apart from carefully selected fields in a production environment). Crafting a regex for this can be flaky. Especially if you need to deal with multiple lines and account for integers, arrays, all sorts of data types. Or if you want to be able to expose some fields on some topics, but not others, etc.

I've been trying my fork locally which introduces a new JSON filtering system (while keeping the regex stuff entirely in place) and it works great, and I'd like to contribute it upstream. I think I've added more than sufficient tests as well, and retained backwards compatibility by defaulting to using the regex filtering if not defined (which current users won't, since `mode` is a new property).

It has two modes - `json_mask_by_default` (where everything is masked by default and you have to specifically opt in to see specific fields - useful for regulation-heavy environments) and `json_show_by_default` where everything shows by default but you can mask specific bits of data.

jamfor352 commented:
Hey @tchiotludo , I just pushed a commit as I realised there were some flaky tests as well. Not strictly related to this change, but a nice to have

AlexisSouquiere commented:
@jamfor352 I will take time to review it and give you my feedback next week

emerfan commented:
Hi, have been using this build locally and it's awesome - following the PR

jamfor352 commented:
> Hi, have been using this build locally and it's awesome - following the PR

Thank you :) Glad to hear it's worked well for you!

AlexisSouquiere commented:
@jamfor352 thanks for the PR it's really interesting 😄 
If you can check the different comments I put. Don't hesitate if you have a question !

jamfor352 commented:
> @jamfor352 thanks for the PR it's really interesting 😄 
> Can you check the different comments I put ? Don't hesitate if you have a question !

Thanks, they all seem reasonable - will get around to it sometime this week.

jamfor352 commented:
Will update in the coming days - been busy moving house so haven't had much free time!

emerfan commented:
Keeping it fresh in minds - I'm very interested in deploying this build, I can obviously pull it down and make the changes myself but would prefer to use the master build

jamfor352 commented on src/main/java/org/akhq/utils/NoOpMasker.java:
```diff
@@ -0,0 +1,11 @@
+
+import org.akhq.models.Record;
+
+public class NoOpMasker implements Masker {
```
@AlexisSouquiere This is the only change that was requested that I adjusted a little - I actually defaulted it to the Regex one for backwards compatibility for current users (ie they won't upgrade, and suddenly their filters stop working). The regex filter currently runs all the time by default anyway. Didn't want this to be a breaking change

jamfor352 commented:
Hi @AlexisSouquiere I've updated my PR based on recommendations (with one small difference, re: default masking filter, to not break current compatibility).
cc: @emerfan

jamfor352 commented:
Hi @AlexisSouquiere I've updated my PR based on recommendations (with one small difference, re: default masking filter, to not break current compatibility). Added documentation, too.

I've also made some other improvements that I've spotted - made the filtering more consistent (ie, always using qualified paths for both JSON maskers), and also some slight performance improvements.

cc: @emerfan could you try this new build out and confirm everything is working okay for you as well after the changes? Please note the change to `json_mask_by_default` where nested objects are now distinct from each other (eg `metadata.value` and `address.value` need separate entries, rather than just `value`) to ensure it still works 👍

jamfor352 commented:
> Hi @jamfor352 this isn't working for me .... debugging I can see that masking is being applied for the record (I'm using `json_mask_by_default`, but the unmasked record is being returned to the UI. Sanity checked by dropping your original try/catch logic in to the `maskRecord` method and that's still working. I'm guessing something funky up with the RecordRepository / how it's returning the record maybe - I'll get back to it later and dig further
> 
> ```
>   security:
>     data-masking:
>       mode: json_mask_by_default
>       json-filters:
>         - description: Unmask non-sensitive values
>           topic: results
>           keys:
>             - Results
> ```

Could you provide an example record for me to try out with my end? All the tests are passing, which is odd based on how the masker is activated. Will take another look.

jamfor352 commented:
> > Hi @jamfor352 this isn't working for me .... debugging I can see that masking is being applied for the record (I'm using `json_mask_by_default`, but the unmasked record is being returned to the UI. Sanity checked by dropping your original try/catch logic in to the `maskRecord` method and that's still working. I'm guessing something funky up with the RecordRepository / how it's returning the record maybe - I'll get back to it later and dig further
> > 
> > ```
> >   security:
> >     data-masking:
> >       mode: json_mask_by_default
> >       json-filters:
> >         - description: Unmask non-sensitive values
> >           topic: results
> >           keys:
> >             - Results
> > ```
> 
> Could you provide an example record for me to try out with my end? All the tests are passing, which is odd based on how the masker is activated. In theory the logic hasnt changed either beyond the nested field adjustment. Will take another look.

emerfan commented:
> > > Hi @jamfor352 this isn't working for me .... debugging I can see that masking is being applied for the record (I'm using `json_mask_by_default`, but the unmasked record is being returned to the UI. Sanity checked by dropping your original try/catch logic in to the `maskRecord` method and that's still working. I'm guessing something funky up with the RecordRepository / how it's returning the record maybe - I'll get back to it later and dig further
> > > ```
> > >   security:
> > >     data-masking:
> > >       mode: json_mask_by_default
> > >       json-filters:
> > >         - description: Unmask non-sensitive values
> > >           topic: results
> > >           keys:
> > >             - Results
> > > ```
> 
> Could you provide an example record for me to try out with my end? All the tests are passing, which is odd based on how the masker is activated. In theory the logic hasnt changed either beyond the nested field adjustment. Will take another look.

Here is a close sample, there's nothing complex about it though... I've tested this value in the unit tests and it also passes.
```
            {
              "Product": "Cheese",
              "Id": 12345,
              "SellByDate": "2024-01-11T00:00:00+00:00",
              "IsDelicious": true
            }
```
Have you tried running it and testing with the UI / Kafka? 

If I replace the updated maskRecord with your original one, it works fine. I can debug to the point I can see the masked record returned to RecordRepository but not as easy to see from there what's happening

```
    public Record maskRecord(Record record) {
        try {
            if(record.getValue().trim().startsWith("{") && record.getValue().trim().endsWith("}")) {
                JsonMaskingFilter foundFilter = null;
                for (JsonMaskingFilter filter : jsonMaskingFilters) {
                    if (record.getTopic().getName().equalsIgnoreCase(filter.getTopic())) {
                        foundFilter = filter;
                    }
                }
                if (foundFilter != null) {
                    return applyMasking(record, foundFilter.getKeys());
                } else {
                    return applyMasking(record, List.of());
                }
            } else {
                return record;
            }
        } catch (Exception e) {
            LOG.error("Error masking record", e);
            return record;
        }
    }
```

jamfor352 commented:
> > > > Hi @jamfor352 this isn't working for me .... debugging I can see that masking is being applied for the record (I'm using `json_mask_by_default`, but the unmasked record is being returned to the UI. Sanity checked by dropping your original try/catch logic in to the `maskRecord` method and that's still working. I'm guessing something funky up with the RecordRepository / how it's returning the record maybe - I'll get back to it later and dig further
> > > > ```
> > > >   security:
> > > >     data-masking:
> > > >       mode: json_mask_by_default
> > > >       json-filters:
> > > >         - description: Unmask non-sensitive values
> > > >           topic: results
> > > >           keys:
> > > >             - Results
> > > > ```
> > 
> > Could you provide an example record for me to try out with my end? All the tests are passing, which is odd based on how the masker is activated. In theory the logic hasnt changed either beyond the nested field adjustment. Will take another look.
> 
> Here is a close sample, there's nothing complex about it though... I've tested this value in the unit tests and it also passes.
> ```
>             {
>               "Product": "Cheese",
>               "Id": 12345,
>               "SellByDate": "2024-01-11T00:00:00+00:00",
>               "IsDelicious": true
>             }
> ```
> Have you tried running it and testing with the UI / Kafka? 
> 
> If I replace the updated maskRecord with your original one, it works fine. I can debug to the point I can see the masked record returned to RecordRepository but not as easy to see from there what's happening
> 
> ```
>     public Record maskRecord(Record record) {
>         try {
>             if(record.getValue().trim().startsWith("{") && record.getValue().trim().endsWith("}")) {
>                 JsonMaskingFilter foundFilter = null;
>                 for (JsonMaskingFilter filter : jsonMaskingFilters) {
>                     if (record.getTopic().getName().equalsIgnoreCase(filter.getTopic())) {
>                         foundFilter = filter;
>                     }
>                 }
>                 if (foundFilter != null) {
>                     return applyMasking(record, foundFilter.getKeys());
>                 } else {
>                     return applyMasking(record, List.of());
>                 }
>             } else {
>                 return record;
>             }
>         } catch (Exception e) {
>             LOG.error("Error masking record", e);
>             return record;
>         }
>     }
> ```
> 

Can you provide your full application.yml, if possible?

emerfan commented:
> > > > > Hi @jamfor352 this isn't working for me .... debugging I can see that masking is being applied for the record (I'm using `json_mask_by_default`, but the unmasked record is being returned to the UI. Sanity checked by dropping your original try/catch logic in to the `maskRecord` method and that's still working. I'm guessing something funky up with the RecordRepository / how it's returning the record maybe - I'll get back to it later and dig further
> > > > > ```
> > > > >   security:
> > > > >     data-masking:
> > > > >       mode: json_mask_by_default
> > > > >       json-filters:
> > > > >         - description: Unmask non-sensitive values
> > > > >           topic: results
> > > > >           keys:
> > > > >             - Results
> > > > > ```
> > > 
> > > 
> > > Could you provide an example record for me to try out with my end? All the tests are passing, which is odd based on how the masker is activated. In theory the logic hasnt changed either beyond the nested field adjustment. Will take another look.
> > 
> > 
> > Here is a close sample, there's nothing complex about it though... I've tested this value in the unit tests and it also passes.
> > ```
> >             {
> >               "Product": "Cheese",
> >               "Id": 12345,
> >               "SellByDate": "2024-01-11T00:00:00+00:00",
> >               "IsDelicious": true
> >             }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Have you tried running it and testing with the UI / Kafka?
> > If I replace the updated maskRecord with your original one, it works fine. I can debug to the point I can see the masked record returned to RecordRepository but not as easy to see from there what's happening
> > ```
> >     public Record maskRecord(Record record) {
> >         try {
> >             if(record.getValue().trim().startsWith("{") && record.getValue().trim().endsWith("}")) {
> >                 JsonMaskingFilter foundFilter = null;
> >                 for (JsonMaskingFilter filter : jsonMaskingFilters) {
> >                     if (record.getTopic().getName().equalsIgnoreCase(filter.getTopic())) {
> >                         foundFilter = filter;
> >                     }
> >                 }
> >                 if (foundFilter != null) {
> >                     return applyMasking(record, foundFilter.getKeys());
> >                 } else {
> >                     return applyMasking(record, List.of());
> >                 }
> >             } else {
> >                 return record;
> >             }
> >         } catch (Exception e) {
> >             LOG.error("Error masking record", e);
> >             return record;
> >         }
> >     }
> > ```
> 
> Can you provide your full application.yml, if possible?

akhq:
  connections:
    local:
      properties:
        bootstrap.servers: "localhost:9092"
  security:
    data-masking:
      mode: json_mask_by_default
      json-filters:
        - description: Unmask non-sensitive values
          topic: results
          keys:
            - Product

jamfor352 commented:
> > > > > > Hi @jamfor352 this isn't working for me .... debugging I can see that masking is being applied for the record (I'm using `json_mask_by_default`, but the unmasked record is being returned to the UI. Sanity checked by dropping your original try/catch logic in to the `maskRecord` method and that's still working. I'm guessing something funky up with the RecordRepository / how it's returning the record maybe - I'll get back to it later and dig further
> > > > > > ```
> > > > > >   security:
> > > > > >     data-masking:
> > > > > >       mode: json_mask_by_default
> > > > > >       json-filters:
> > > > > >         - description: Unmask non-sensitive values
> > > > > >           topic: results
> > > > > >           keys:
> > > > > >             - Results
> > > > > > ```
> > > > 
> > > > 
> > > > Could you provide an example record for me to try out with my end? All the tests are passing, which is odd based on how the masker is activated. In theory the logic hasnt changed either beyond the nested field adjustment. Will take another look.
> > > 
> > > 
> > > Here is a close sample, there's nothing complex about it though... I've tested this value in the unit tests and it also passes.
> > > ```
> > >             {
> > >               "Product": "Cheese",
> > >               "Id": 12345,
> > >               "SellByDate": "2024-01-11T00:00:00+00:00",
> > >               "IsDelicious": true
> > >             }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Have you tried running it and testing with the UI / Kafka?
> > > If I replace the updated maskRecord with your original one, it works fine. I can debug to the point I can see the masked record returned to RecordRepository but not as easy to see from there what's happening
> > > ```
> > >     public Record maskRecord(Record record) {
> > >         try {
> > >             if(record.getValue().trim().startsWith("{") && record.getValue().trim().endsWith("}")) {
> > >                 JsonMaskingFilter foundFilter = null;
> > >                 for (JsonMaskingFilter filter : jsonMaskingFilters) {
> > >                     if (record.getTopic().getName().equalsIgnoreCase(filter.getTopic())) {
> > >                         foundFilter = filter;
> > >                     }
> > >                 }
> > >                 if (foundFilter != null) {
> > >                     return applyMasking(record, foundFilter.getKeys());
> > >                 } else {
> > >                     return applyMasking(record, List.of());
> > >                 }
> > >             } else {
> > >                 return record;
> > >             }
> > >         } catch (Exception e) {
> > >             LOG.error("Error masking record", e);
> > >             return record;
> > >         }
> > >     }
> > > ```
> > 
> > Can you provide your full application.yml, if possible?
> 
> ```
> akhq:
>   connections:
>     local:
>       properties:
>         bootstrap.servers: "localhost:9092"
>   security:
>     data-masking:
>       mode: json_mask_by_default
>       json-filters:
>         - description: Unmask non-sensitive values
>           topic: results
>           keys:
>             - Product
> ```
> 
> 

I'm wondering if it's something to do with the "disable if no filters found" and there being a bug with determining this

Thanks for the YAML - I'll check it out later.

emerfan commented:
> > > > > > > Hi @jamfor352 this isn't working for me .... debugging I can see that masking is being applied for the record (I'm using `json_mask_by_default`, but the unmasked record is being returned to the UI. Sanity checked by dropping your original try/catch logic in to the `maskRecord` method and that's still working. I'm guessing something funky up with the RecordRepository / how it's returning the record maybe - I'll get back to it later and dig further
> > > > > > > ```
> > > > > > >   security:
> > > > > > >     data-masking:
> > > > > > >       mode: json_mask_by_default
> > > > > > >       json-filters:
> > > > > > >         - description: Unmask non-sensitive values
> > > > > > >           topic: results
> > > > > > >           keys:
> > > > > > >             - Results
> > > > > > > ```
> > > > > 
> > > > > 
> > > > > Could you provide an example record for me to try out with my end? All the tests are passing, which is odd based on how the masker is activated. In theory the logic hasnt changed either beyond the nested field adjustment. Will take another look.
> > > > 
> > > > 
> > > > Here is a close sample, there's nothing complex about it though... I've tested this value in the unit tests and it also passes.
> > > > ```
> > > >             {
> > > >               "Product": "Cheese",
> > > >               "Id": 12345,
> > > >               "SellByDate": "2024-01-11T00:00:00+00:00",
> > > >               "IsDelicious": true
> > > >             }
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > Have you tried running it and testing with the UI / Kafka?
> > > > If I replace the updated maskRecord with your original one, it works fine. I can debug to the point I can see the masked record returned to RecordRepository but not as easy to see from there what's happening
> > > > ```
> > > >     public Record maskRecord(Record record) {
> > > >         try {
> > > >             if(record.getValue().trim().startsWith("{") && record.getValue().trim().endsWith("}")) {
> > > >                 JsonMaskingFilter foundFilter = null;
> > > >                 for (JsonMaskingFilter filter : jsonMaskingFilters) {
> > > >                     if (record.getTopic().getName().equalsIgnoreCase(filter.getTopic())) {
> > > >                         foundFilter = filter;
> > > >                     }
> > > >                 }
> > > >                 if (foundFilter != null) {
> > > >                     return applyMasking(record, foundFilter.getKeys());
> > > >                 } else {
> > > >                     return applyMasking(record, List.of());
> > > >                 }
> > > >             } else {
> > > >                 return record;
> > > >             }
> > > >         } catch (Exception e) {
> > > >             LOG.error("Error masking record", e);
> > > >             return record;
> > > >         }
> > > >     }
> > > > ```
> > > 
> > > 
> > > Can you provide your full application.yml, if possible?
> > 
> > 
> > ```
> > akhq:
> >   connections:
> >     local:
> >       properties:
> >         bootstrap.servers: "localhost:9092"
> >   security:
> >     data-masking:
> >       mode: json_mask_by_default
> >       json-filters:
> >         - description: Unmask non-sensitive values
> >           topic: results
> >           keys:
> >             - Product
> > ```
> 
> I'm wondering if it's something to do with the "disable if no filters found" (which I have realised probably shouldnt be in the mask by default mode) nd there being a bug with determining this, or perhaps the JSON parsing
> 
> Thanks for the YAML - I'll check it out later.

Just to make this slightly weirder .... 

For easier testing, I created a topic and published a single message to it. 

When I run the application, the record is returned unmasked. When I debug through the masker, the record is eventually returned masked as expected ... 😕

jamfor352 commented:
> Just to make this slightly weirder ....
> 
> For easier testing, I created a topic and published a single message to it.
> 
> When I run the application, the record is returned unmasked. When I run in debug mode and debug through the masker, the record is eventually returned (to the UI) masked as expected ... 😕

I've just pushed a few code changes. I'm unsure what the issue you're finding is but yeah, can you try out the new build?

jamfor352 commented:
> Just to make this slightly weirder ....
> 
> For easier testing, I created a topic and published a single message to it.
> 
> When I run the application, the record is returned unmasked. When I run in debug mode and debug through the masker, the record is eventually returned (to the UI) masked as expected ... 😕

I'm not entirely sure what the issue here is... but I've changed a few things relating to how masking is validated to work (removed the `enabled` functionality since it doesn't really bring much and breaks the mask by default flow, and switched JSON checking to a different library which may be better suited)

emerfan commented:
> > Just to make this slightly weirder ....
> > For easier testing, I created a topic and published a single message to it.
> > When I run the application, the record is returned unmasked. When I run in debug mode and debug through the masker, the record is eventually returned (to the UI) masked as expected ... 😕
> 
> I'm not entirely sure what the issue here is... but I've changed a few things relating to how masking is validated to work (removed the `enabled` functionality since it doesn't really bring much and breaks the mask by default flow, and switched JSON checking to a different library which may be better suited)

Does this work as expected for you against a cluster? I'm still seeing the debug vs run issue

emerfan commented:
> > > Just to make this slightly weirder ....
> > > For easier testing, I created a topic and published a single message to it.
> > > When I run the application, the record is returned unmasked. When I run in debug mode and debug through the masker, the record is eventually returned (to the UI) masked as expected ... 😕
> > 
> > 
> > I'm not entirely sure what the issue here is... but I've changed a few things relating to how masking is validated to work (removed the `enabled` functionality since it doesn't really bring much and breaks the mask by default flow, and switched JSON checking to a different library which may be better suited)
> 
> Does this work as expected for you against a cluster? I'm still seeing the debug vs run issue

I've found the issue ... it's with the new isTombstone and isJson methods. The record is initially null when it hits them. 

This works:

```
public interface Masker {

    Logger LOG = LoggerFactory.getLogger(Masker.class);

    Record maskRecord(Record record);

    default boolean isTombstone(Record record) {
        return record.getValue() == null;
    }

   default boolean isJson(Record record) {
        if(isTombstone(record)) {
            return false;
        }
        try {
            new JSONObject(record.getValue());
        } catch (JSONException ex) {
            try {
                new JSONArray(record.getValue());
            } catch (JSONException ex1) {
                return false;
            }
        }
        return true;
    }
}
```


and 

```
    public Record maskRecord(Record record) {
        try {
            if(isTombstone(record)) {
                LOG.debug("Record at topic {}, partition {}, offset {} is a tombstone, so not masking.", record.getTopic(), record.getPartition(), record.getOffset());
                return record;
            } else if(isJson(record)) {
                return jsonMaskingFilters
                    .stream()
                    .filter(jsonMaskingFilter -> record.getTopic().getName().equalsIgnoreCase(jsonMaskingFilter.getTopic()))
                    .findFirst()
                    .map(filter -> applyMasking(record, filter.getKeys()))
                    .orElseGet(() -> applyMasking(record, List.of()));
            } else {
                LOG.debug("Record at topic {}, partition {}, offset {} is not JSON, so not masking.", record.getTopic(), record.getPartition(), record.getOffset());
                return record;
            }
        } catch (Exception e) {
            LOG.error("Error masking record at topic {}, partition {}, offset {} due to {}", record.getTopic(), record.getPartition(), record.getOffset(), e.getMessage());
            record.setValue("An exception occurred during an attempt to mask this record. This record is unavailable to view due to safety measures from json_mask_by_default to not leak sensitive data. Please contact akhq administrator.");
            return record;
        }
    }
```

jamfor352 commented:
> > > > Just to make this slightly weirder ....
> > > > For easier testing, I created a topic and published a single message to it.
> > > > When I run the application, the record is returned unmasked. When I run in debug mode and debug through the masker, the record is eventually returned (to the UI) masked as expected ... 😕
> > > 
> > > 
> > > I'm not entirely sure what the issue here is... but I've changed a few things relating to how masking is validated to work (removed the `enabled` functionality since it doesn't really bring much and breaks the mask by default flow, and switched JSON checking to a different library which may be better suited)
> > 
> > 
> > Does this work as expected for you against a cluster? I'm still seeing the debug vs run issue
> 
> I've found the issue ... it's with the new isTombstone and isJson methods. The record is initially null when it hits them.
> 
> This works:
> 
> ```
> public interface Masker {
> 
>     Logger LOG = LoggerFactory.getLogger(Masker.class);
> 
>     Record maskRecord(Record record);
> 
>     default boolean isTombstone(Record record) {
>         return record.getValue() == null;
>     }
> 
>    default boolean isJson(Record record) {
>         if(isTombstone(record)) {
>             return false;
>         }
>         try {
>             new JSONObject(record.getValue());
>         } catch (JSONException ex) {
>             try {
>                 new JSONArray(record.getValue());
>             } catch (JSONException ex1) {
>                 return false;
>             }
>         }
>         return true;
>     }
> }
> ```
> 
> and
> 
> ```
>     public Record maskRecord(Record record) {
>         try {
>             if(isTombstone(record)) {
>                 LOG.debug("Record at topic {}, partition {}, offset {} is a tombstone, so not masking.", record.getTopic(), record.getPartition(), record.getOffset());
>                 return record;
>             } else if(isJson(record)) {
>                 return jsonMaskingFilters
>                     .stream()
>                     .filter(jsonMaskingFilter -> record.getTopic().getName().equalsIgnoreCase(jsonMaskingFilter.getTopic()))
>                     .findFirst()
>                     .map(filter -> applyMasking(record, filter.getKeys()))
>                     .orElseGet(() -> applyMasking(record, List.of()));
>             } else {
>                 LOG.debug("Record at topic {}, partition {}, offset {} is not JSON, so not masking.", record.getTopic(), record.getPartition(), record.getOffset());
>                 return record;
>             }
>         } catch (Exception e) {
>             LOG.error("Error masking record at topic {}, partition {}, offset {} due to {}", record.getTopic(), record.getPartition(), record.getOffset(), e.getMessage());
>             record.setValue("An exception occurred during an attempt to mask this record. This record is unavailable to view due to safety measures from json_mask_by_default to not leak sensitive data. Please contact akhq administrator.");
>             return record;
>         }
>     }
> ```

Good catch! I put this in the domain object as I thought it would be cleaner.
I'll update my branch with this tonight

emerfan commented:
> @emerfan updated my branch, so it should all be working out of the box for you now.
> 
> @AlexisSouquiere can I get a re-review please now changes have been applied and bugs fixed? 🙏

Looks great! I've run it against a substantial cluster. I like the mask all topics by default, and the updated handling of nested objects. This is a great feature, thanks!

jamesforwardnwboxed commented:
Hey @AlexisSouquiere just tagging so this doesn't get lost. Could I get a re-review please

jamfor352 commented:
> @jamfor352 I will review it tomorrow !

Thank you!

AlexisSouquiere reviewed (changes requested):
Few updates needed and then it's good for me

jamfor352 commented:
Hey @AlexisSouquiere applied all the changes requested. Although I also realised I may as well inline the `isTombstone` logic into the `isJson` since it was no longer worth having the `isTombstone` method anyway since it's never used elsewhere and the logical conclusion only really makes sense here.

Hope this looks all OK now 👍

jamfor352 commented on src/main/java/org/akhq/utils/RegexMasker.java:
```diff
@@ -0,0 +1,42 @@
+
+    public RegexMasker(DataMasking  dataMasking ) {
+        this.filters = dataMasking.getFilters();
+    }    @Override
```
Fixed

jamfor352 commented on src/main/java/org/akhq/utils/JsonShowByDefaultMasker.java:
```diff
@@ -0,0 +1,69 @@
+        this.jsonMaskReplacement = dataMasking.getJsonMaskReplacement();
+    }
+
+    public Record maskRecord(Record record) {
```
Fixed (just with the isTombstone inlined into isJson as it's only used there and only really makes sense here)

jamfor352 commented on src/main/java/org/akhq/models/Record.java:
```diff
@@ -189,6 +194,10 @@ public void setTruncated(Boolean truncated) {
+    public void setTopic(Topic topic) {
+        this.topic = topic;
+    }
+
```
Fixed

jamfor352 commented on src/main/java/org/akhq/models/Record.java:
```diff
@@ -26,6 +28,9 @@
 import org.apache.kafka.common.serialization.Deserializer;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
```
Fixed

jamfor352 commented on src/main/java/org/akhq/models/Record.java:
```diff
@@ -5,6 +5,8 @@
 import com.fasterxml.jackson.annotation.JsonView;
 import com.fasterxml.jackson.databind.JsonNode;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonSyntaxException;
```
Fixed

jamfor352 commented on src/main/java/org/akhq/configs/DataMasking.java:
```diff
@@ -9,5 +9,8 @@
+   List<RegexFilter> filters = new ArrayList<>(); // regex filters still use `filters` for backwards compatibility
+   List<JsonMaskingFilter> jsonFilters = new ArrayList<>();
+   String jsonMaskReplacement = "xxxx";
+   boolean cachingEnabled = false;
```
Fixed

AlexisSouquiere reviewed (approved):
Sounds good to me !

jamfor352 commented:
> Sounds good to me !

Thanks for the approval! What is the process in terms of getting this upstream, does it need more reviewers?

tchiotludo commented:
@jamfor352 nothing required, thanks for the contributions :clap:
